src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerationsHelpers.cs

```diff
@@ -186,116 +205,174 @@ private static bool IsInvocationCausingEnumerationOverInvocationInstance(IInvoca
+            {
+                return IsLinqChainInvocation(invocationOperation, argumentOperation, wellKnownSymbolsInfo, out enumerateArgument);
+            }
+
+            enumerateArgument = false;
+            return false;
+        }
 
         /// <summary>
         /// Check if <param name="argumentOperationToCheck"/> is passed as a deferred executing argument into <param name="invocationOperation"/>.
         /// </summary>
-        public static bool IsDeferredExecutingInvocation(
+        public static bool IsLinqChainInvocation(
             IInvocationOperation invocationOperation,
             IArgumentOperation argumentOperationToCheck,
-            WellKnownSymbolsInfo wellKnownSymbolsInfo)
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            out bool enumerateArgument)
         {
+            enumerateArgument = false;
             RoslynDebug.Assert(invocationOperation.Arguments.Contains(argumentOperationToCheck));
             if (!IsDeferredType(argumentOperationToCheck.Value.Type?.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
             {
                 return false;
             }
 
             var targetMethod = invocationOperation.TargetMethod;
+            // For C#, extension method is used as an ordinary static method.
+            // For VB, ex: a.Concat(b)
+            // 'b' is an argument to 'Concat', which is a reduced method.
             var reducedFromMethod = targetMethod.ReducedFrom ?? targetMethod;
+            var originalMethod = reducedFromMethod.OriginalDefinition;
+            if (!IsDeferredType(targetMethod.ReturnType.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                return false;
+            }
+
+            // For VB, ex: a.Concat(b)
+            // 'b' is in fact the first argument to 'Concat', because the extension method in VB is reduced.
             var argumentMappingParameter = targetMethod.MethodKind == MethodKind.ReducedExtension
                 ? GetReducedFromParameter(targetMethod, argumentOperationToCheck.Parameter)
                 : argumentOperationToCheck.Parameter;
 
-            if (wellKnownSymbolsInfo.DeferredMethods.Contains(reducedFromMethod.OriginalDefinition)
-                && reducedFromMethod.Parameters.Any(
-                    methodParameter => IsDeferredType(methodParameter.Type?.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes) && methodParameter.Equals(argumentMappingParameter)))
+            if (wellKnownSymbolsInfo.LinqChainMethods.Contains(originalMethod)
+                && originalMethod.Parameters.Any(
+                    methodParameter => IsDeferredType(methodParameter.Type?.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes) && methodParameter.Equals(argumentMappingParameter.OriginalDefinition)))
             {
+                // All well-known linq chain method under Linq namespace won't enumerated the argument.
```

Comment:
NIT: isn't it should return false if it `won't enumerate`?
```suggestion
                // All well-known linq chain methods under Linq namespace won't enumerate the argument.
```